### ISSUE TITLE
[clang] Miscompile with -O3 and -O0/1/2 since 18.1.0

### ISSUE THREAD

**Issue by cardigan1008** [llvm/llvm-project#115149]
When I compiled this code with -O0/1/2, its output is 0. However, with -O3, it returned 143:

```c
int printf(const char *, ...);
char a, b;
int c;
char *e = &b;
int f(char *g, int *k) {
  char *d = g + *k;
  for (; *d && *d <= ' '; d++)
    ;
  if (*d)
    return 0;
  return 1;
}
int l(int g) {
  char h[] = {a, a, a};
  int i[] = {g};
  int j = f(h, i);
  return j;
}
long m() {
  *e = 255;
  for (; l(b + 1);)
    return 0;
  for (;;)
    ;
}
int main() {
  m();
  printf("%d\n", c);
}
```

Details can be found here: https://godbolt.org/z/dnvvTPror

**Comment by antoniofrighetto** [comment]
This looks like a nasty one. Getting the optimized IR via -O3 and setting noinline/optnone to main, we get: https://llvm.godbolt.org/z/vaPvs6d16. 

This is the initial assembly output with -O1 (which leads to the infinite loop):

```asm
main:                                   # @main
        push    rax
        mov     rax, qword ptr [e]
        mov     byte ptr [rax], -1
        movsx   rax, byte ptr [b]
        mov     cl, byte ptr [a]
        mov     byte ptr [rsp + 6], cl
        mov     byte ptr [rsp + 7], cl
        lea     rdx, [rsp + 5]
        add     rdx, rax
        mov     rax, rdx
.LBB0_1:                                # %for.cond.i.i.i
        add     rax, 1
        mov     cl, byte ptr [rdx + 1]
        ...
```
Whereas this is with -O0 (returns 0):
```asm
main:                                   # @main
        sub     rsp, 24
        mov     rax, qword ptr [e]
        mov     byte ptr [rax], -1
        movsx   rcx, byte ptr [b]
        mov     al, byte ptr [a]
        mov     byte ptr [rsp + 22], al
        mov     byte ptr [rsp + 23], al
        lea     rax, [rsp + 21]
        add     rax, rcx
        mov     qword ptr [rsp + 8], rax        # 8-byte Spill
.LBB0_1:                                # %for.cond.i.i.i
        mov     rcx, qword ptr [rsp + 8]        # 8-byte Reload
        mov     rax, rcx
        add     rax, 1
        mov     dl, byte ptr [rcx + 1]
```
In the optimized one, $rdx is set to point to $rsp+0x5, part of the stack slot containing $rax (undef), later dereferenced to set the condition leading to the loop. Initially I thought some bug in regalloc, but to me it seems more something in prologepiloginserter. Function prologue / epilogue seems set up correctly with triple arm64-apple-macosx.

**Comment by antoniofrighetto** [comment]
Looking at the IR better, it seems like InstCombine performs a wrong optimization immediately after inlining `l` into `m`, probably related with default argument conversion done by operator + (which should include integer promotion on types smaller than integer). Alive2 seems to confirm it: https://alive2.llvm.org/ce/z/w7iAkR.

**Comment by AreaZR** [comment]
Can we backport this to 19.x?

**Comment by antoniofrighetto** [comment]
Should be fixed in 929cbe7f596733f85cd274485acc19442dd34a80, thanks for reporting!

**Comment by AreaZR** [comment]
@antoniofrighetto xan we please backport this
